How-To for Noise Models

obliviateandsurrender opened the pull request:
**Title:** How to use noise models in PennyLane

**Summary:** Elaborates on constructing and using noise models in PennyLane via the new `noise` module.

**Relevant references:**

**Possible Drawbacks:**

**Related GitHub Issues:**

----
If you are writing a demonstration, please answer these questions to facilitate the marketing process.

* GOALS — Why are we working on this now?  --> Promote a new PL feature

* AUDIENCE — Who is this for? --> Researchers and users of PennyLane

* KEYWORDS — What words should be included in the marketing post? --> Noise models, noisy simulation

* Which of the following types of documentation is most similar to your file? 
(more details [here](https://www.notion.so/xanaduai/Different-kinds-of-documentation-69200645fe59442991c71f9e7d8a77f8))
    
- [ ] Tutorial
- [ ] Demo
- [X] How-to

trbromley commented on _static/demonstration_assets/noise_models/noise_model_long.jpg:
Nice! How challenging would this be to get to work with our default `draw_mpl` style?

obliviateandsurrender replied:
![image](https://github.com/user-attachments/assets/5960d20f-ad06-4996-9292-bffe711b5653)
It's hard to read damping error boxes in the default `draw_mpl` style. That's why I had to make the circuits manually.

obliviateandsurrender replied:
I think it would be better to do it later for all the error channels in a single concentrated effort. 🤔

obliviateandsurrender replied:
I would prefer to do this later with a fix for all error channels with a more detailed discussion rather than a quick fix-PR.

trbromley commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,360 @@
+# evolution of and inference from quantum states under effect of state preparation, measurement and
+# environment-based errors.
+#
+# Here, we introduce how to use the PennyLane :mod:`~pennylane.noise` module for constructing and
```
The link isn't working

trbromley replied:
```suggestion
# Here, we introduce how to use the PennyLane :mod:`~.pennylane.noise` module for constructing and
```

trbromley commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,360 @@
+#
+
+######################################################################
+# Let’s understand this briefly with help of the following noise model that inserts an ampltitude and
```
```suggestion
# Let’s understand this briefly with help of the following noise model that inserts amplitude and
```

trbromley commented on _static/demonstration_assets/noise_models/noise_model_short.jpg:
There's a typo in the left blue box.

trbromley commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,360 @@
+# Boolean functions
+# ~~~~~~~~~~~~~~~~~
+#
+# We want to implement conditions as boolean functions called ``Conditionals`` that accepts
```
```suggestion
# We want to implement conditions as Boolean functions called ``Conditionals`` that accepts
```

trbromley replied:
Do we need the double backticks around ``Conditionals``? I don't think it's a programmatic quantity we've introduced in PL, more of a general concept.

soranjh reviewed (changes requested):
Thanks @obliviateandsurrender, looks good to me overall. Left some corrections and comments.

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,372 @@
+# the corresponding callable would queue the noise operations for the evaluated gate based on the
+# user-provided metadata such as hardware topology constraints and relaxation/dephasing times.
+#
+# Let’s look at an example for a noise model that implements these main components.
+# This model inserts amplitude and phase damping errors for :class:`~.pennylane.RX`
+# and :class:`~.pennylane.RY` gates, respectively. So, the conditions in the model
+# are simple checks for verifying if a circuit gate is any of :class:`~.pennylane.RX`
+# and :class:`~.pennylane.RY` gates and the noise operations are amplitude and phase
+# damping errors with their damping probabilities as some function of rotation angles.
+# The following image shows how the model transforms a sample circuit by inserting
+# the appropriate error operation after each gate that evaluates positively for the
+# model conditions.
```
```suggestion
# The following example shows how a noise model transforms a sample circuit by inserting
# amplitude and phase damping errors for :class:`~.pennylane.RX`
# and :class:`~.pennylane.RY` gates, respectively.
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,372 @@
+#
+# Here, we show how to use the features provided in the :mod:`~pennylane.noise` module of PennyLane
+# to construct and manipulate noise models for enabling noisy simulation. PennyLane supports
+# `insertion`-based noise models. These models are constructed from three main components:
+# conditions for applying noise operations, callables that apply user-defined noise operations and
+# optionally noise-related metadata. Each condition evaluates gate operations in the quantum
```
```suggestion
# `insertion`-based noise models. These models are constructed from two main components:
# conditions for applying noise operations and callables that apply user-defined noise operations with
# optional noise-related metadata. Each condition evaluates gate operations in the quantum
```

isaacdevlugt commented on demonstrations/tutorial_noise_models.metadata.json:
```diff
@@ -0,0 +1,35 @@
+            "uri": "/_static/demonstration_assets/regular_demo_thumbnails/thumbnail_optimizing_noisy_circuits_Cirq.png"
+        }
+    ],
+    "seoDescription": "Learn how noise models can be built and attached to a quantum circuit in PennyLane.",
```
Not sure if "attached" is the right word 🤔

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,366 @@
+# ``True`` for :math:`R_X(\phi)` gate operations with :math:`\phi < 1.0` can be constructed as:
+#
+
+
```
I have to turn off auto black-formatting when I write a demo to avoid the extra spacing 😅. Super annoying 😢

isaacdevlugt reviewed (changes requested):
Nice work @obliviateandsurrender! The examples you chose really show off the features well, both technically and UI'ly (? — I just made up that word) 💪. 

Overall, more brevity (where possible) would serve this how-to well, since we just want the reader to be able to absorb the essential information and have links to elsewhere in the documentation if they require more detail. Right now, the draft reads more like a demo than a how-to 🧠. Brevity here can be achieved in a few ways:

- Itemized lists instead of dedicated sections
- Saying "check out the X for more detailed information" when you think things might be getting too complicated
- Listing to the reader the set of methods / functions to use, but only showing how to use one function that belongs to the set rather than showing them all

I'm definitely not saying that we need to completely restructure your how-to! It might be that some things can be cut, but I like your code examples and how you made natural transitions into subsequent sections. Just needs a polish and I think this will be nice! 😌 

Also, please make sure to fill in the marketing and design briefs for Tarik and Ashley 🎨 🤹

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,366 @@
+
+import pennylane as qml
+
+cond1 = qml.noise.op_eq("X")
+cond2 = qml.noise.op_in(["X", qml.Y, qml.CNOT([0, 1])])
```
Yep, this is because 
![image](https://github.com/user-attachments/assets/fc37ab83-c77e-48db-b8c3-ddda913cbafa)

obliviateandsurrender replied:
1. I will change these to `RX` and `RY` if you are worried about this inconsistency.
2. Regarding the `reprs`, I am just trying to show the different ways conditionals can be constructed.

isaacdevlugt replied:
Thanks Utkarsh! I think the least confusing thing for the reader will be to use classes 👍

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,366 @@
+# for constructing the wire set:
+#
+
+cond3 = qml.noise.wires_eq("aux")
```
That's because `aux` is a single element and `c` appears within a `list`.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,366 @@
+def thermal_func(op, **kwargs):
+    qml.ThermalRelaxationError(0.4, kwargs["t1"], 0.2, 0.6, op.wires)
+
+######################################################################
+# Now to see what error would this noise-function would queue (insert) in the circuit
+# using some example gate operations:
+#
+
+for op, t1 in [(qml.X(0), 0.01), (qml.RZ(1.234, "w1"), 0.02), (qml.S("aux"), 0.03)]:
+    with qml.queuing.AnnotatedQueue() as q:
+        thermal_func(op, t1=t1)
+    print(f"Error for {op}:\n{q.queue[0]}\n")
```
I simplified both of them a bit.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,366 @@
+# We can then use these for running noisy simulations as shown below:
+#
+
+import numpy as np
+
+num_shots = 100000
+ideal_circuit_counts = qcircuit(params, shots=num_shots)
+noisy_circuit_counts = noisy_circuit(params, shots=num_shots)
+noist_dev_counts = noisy_dev_circuit(params, shots=num_shots)
+categories = list(noisy_circuit_counts.keys())
+
+width = 0.2
+bars1 = np.arange(len(categories))
+bars2 = bars1 + width
+bars3 = bars1 + 2 * width
+
+counts_ideal = np.array(list(ideal_circuit_counts.values())) / num_shots
+counts_ncirc = np.array(list(noisy_circuit_counts.values())) / num_shots
+counts_ndevc = np.array(list(noist_dev_counts.values())) / num_shots
+
+# Create the bar plot
+plt.figure(figsize=(10, 6))
+plt.bar(bars1, counts_ideal, width, label="Ideal Results")
+plt.bar(bars2, counts_ncirc, width, label="Noisy Circuit")
+plt.bar(bars3, counts_ndevc, width, label="Noisy Device")
+
+# Add labels, title, and legend
+plt.xlabel("Bitstring")
+plt.ylabel("Probabilities")
+plt.xticks(bars1 + width / 2, categories)
+plt.legend()
+
+# Show the plot
+plt.show()
```
Noisy outputs for no-noise and the noisy device will be the same. That's why I wanted to show the example to showcase the difference between no-noise and noise cases & similarity between noisy qfunc and noisy device cases more explicitly.

isaacdevlugt commented on _static/demonstration_assets/noise_models/noise_model_long.jpg:
Thanks @obliviateandsurrender! This looks great 👍

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,348 @@
+# in a quantum computation, as they allow for simulating the imperfections in state evolution
+# arising from environment-based errors, state preparation routines, measurements, and more.
+#
+# Here, we show how to use the features provided in the PennyLane's :mod:`~.pennylane.noise`
```
Yep, I guess it doesn't work just in the deployment 🤔

josh146 replied:
https://xanaduhq.slack.com/archives/C03BY2J49HT/p1718880936544019

isaacdevlugt commented on demonstrations/tutorial_noise_models.metadata.json:
```diff
@@ -0,0 +1,35 @@
+            "id": "utkarsh_azad"
+        }
+    ],
+    "dateOfPublication": "2024-07-31T00:00:00+00:00",
+    "dateOfLastModification": "2024-07-31T00:00:00+00:00",
```
@obliviateandsurrender let's coordinate with @ikurecic to get a tentative publication date set 😁

ikurecic replied:
We can keep this as a placeholder and set the correct one as a last step, once the review is done.

isaacdevlugt commented on _static/demonstration_assets/regular_demo_thumbnails/thumbnail_noise_models.jpg:
Just putting a note here that this should be replaced with a drawing from Tarik 🧑‍🎨

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,334 @@
+#
+
+######################################################################
+# Conditionals
```
I think we can restructure this section so that it's leading into the example in the `Noise Models` section quicker. Instead of having an section + example for each type of conditional, we can do:

> To create a noise model, we need to define the conditions for which desired noise is present. PennyLane's insertion-based noise model convention lends itself well to this; when a condition is satisfied, insert noise. In PennyLane's noise module, there are built-in ways to create conditions that cover a lot of ground, including:
> 
> - `feature_name1`: short explanation
> - `feature_name2`: short explanation
> ...
> 
> For example, here's how we would define [an example condition]:
> 
> ```
> # code
> ```
> 
> [then we can briefly explain custom conditions]. Here is an example of [a custom condition]:
> 
> ```
> # code
> ```
> 
> You can also combine conditions with &, |, ... [no need to show an example here, unless it's short!]
> 
> [next section]

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,334 @@
+    print(f"Result for {op}: {and_cond(op)}")
+
+######################################################################
+# Callables
```
```suggestion
# Noisy operations
```

We can restructure this section, as well. Maybe something like this:

> With conditions defined, now we can attach noisy operations to those conditions. In PennyLane, there are two ways to do define noisy operations:
> 
> - `partial_wires`: brief explanation
> - custom noise ops: brief explanation
> 
> [a couple examples that show the two features in action ^]
> 
> ```
> # code
> ```
> 
> # [new section] Creating a noise model
> 
> We can now create a `NoiseModel` by stitching together the conditions with noisy operations that are inserted into the circuit if the conditions are satisfied ...
> 
> ```
> # code
> ```

obliviateandsurrender replied:
Had done it but got undoed somehow, fixed it again.

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,334 @@
+print(noise_model)
+
+######################################################################
+# Adding noise models to your workflow
```
With the above sections restructured, it should flow nicely into this section! Just want to demonstrate the `add_noise` transform.

isaacdevlugt reviewed (changes requested):
@obliviateandsurrender I think the sections leading into `Adding noise models to your workflow` could be restructured. See my comments 🙂

isaacdevlugt reviewed (commented):
Nice @obliviateandsurrender! This flows quite nicely and the structure is definitely what we have in mind for a how-to. From here, there isn't a ton of work to be done 🎉!

- Let's use a `partial_wires` example in the `Callables` section, 
- The conditional `cond_fn` and the noisy operation you define in the `Callables` section can be used in the `NoiseModel` you create later on to save on scroll time
- We can limit the `NoiseModel` dictionary to 3 pairs of conditions/callables for brevity
  - `cond_fn` + `thermal_func` (maybe rename them to `fcond1` and `noise1`)
  - `fcond2` + `noise2`
  - `fcond3` + `noise3`
- The SWAP test is a cool example, but a simpler example that just illustrates the concept of applying noise might be better here. It can be a random circuit — no real "application", per se, but enough to showcase how and when noise is inserted in a circuit. 
- We don't have to show how `device` can be transformed (transforming a `QNode` suffices)

Let me know if you have any questions!

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,274 @@
+#    :func:`~pennylane.noise.partial_wires`, which builds a partially evaluated function
+#    based on the given operation with all its arguments frozen except ``wires``.
+# 2. **User-defined noise functions:** For adding more sophesticated noise, one can define
+#    their own quantum function with the signature specified above. This way, one can also
+#    specify their own custom order for inserting the noise by queing the operation being
+#    evaluated via :func:`~pennylane.apply` within the function definition.
```
I think since we don't have an example it is more useful to keep it.

obliviateandsurrender replied:
I meant as an `immediate` example like the one I'd before so this key info doesn't get lost. I'm shifting it to later now as requested.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,273 @@
+#
+
+noisy_circuit = qml.add_noise(swap_circuit, noise_model)
+print(qml.draw(noisy_circuit, decimals=3, max_length=75)(0.2, 0.3))
```
Made it to print in without the messiness.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.metadata.json:
```diff
@@ -0,0 +1,35 @@
+    "basedOnPapers": [],
+    "referencedByPapers": [],
+    "relatedContent": [
+        {
+            "type": "demonstration",
+            "id": "tutorial_noisy_circuits",
+            "weight": 1.0
+        }
```
Not sure. Only one that comes to my head is `Error mitigation with Mitiq` but that would be more suitable to the Qiskit noise models how-to.

ikurecic replied:
They are relevant to various things you mention in the introduction. Seeing as these have to be added manually, you'd be better off having more than less, but it's your decision. It would be good if you didn't resolve comments before I can address them. Thanks.

obliviateandsurrender commented:
Hey @isaacdevlugt I have incorporated the comments and changed the example to have a more generic circuit! However, now that we are re-using the examples to construct one of the condition-callable pairs, I would still like to keep all four pairs.

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,249 @@
+#    be ``cond_fn(operation: Operation) -> bool``.
+#
+# For example, here's how we would define a conditional that checks for :math:`R_X(\phi)`
+# gate operations with :math:`|\phi| < 1.0` and wires :math:`\in \{1, 2\}`:
```
```suggestion
# gate operations with :math:`|\phi| < 1.0` and wires :math:`\in \{0, 1\}`:
```

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,247 @@
+fcond1, noise1 = rx_and_wires_cond, depol_error
+
+######################################################################
+# Next, we construct a pair to mimic the thermal relaxation errors that are
```
```suggestion
# Next, we construct a pair to mimic thermal relaxation errors that are
```

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,247 @@
+# multiple condition-callable pairs, where noisy noise operations are inserted into the
+# circuit when their corresponding given condition is satisfied. For the first pair, we
+# will use the previously constructed conditional and callable to insert a depolarization
+# error for the :class:`~.pennylane.RX` gates that act on the wires :math:`\in \{0, 1\}`.
```
```suggestion
# error after :class:`~.pennylane.RX` gates that satisfy :math:`|\phi| < 1.0` and that act on wires :math:`\in \{0, 1\}`.
```

isaacdevlugt reviewed (commented):
Nice @obliviateandsurrender! The structure is looking great, now we're just honing in on details.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,249 @@
+# By default, noise operations will be inserted *after* the gate that the noise function
+# is evaluated over. However, we can circumvent this! For example, we can add a sandwiching
+# constant-valued rotation errors for :class:`~.pennylane.Hadamard` gates on the wires
+# :math:`\in \{0, 1\}`:
```
I have added this but with some modifications. The main motivation is that I want to keep the term "evaluation" reserved for conditionals and not mix them with noise functions. Otherwise, it may just confuse a reader about their respective usage.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,247 @@
+depol_error = qml.noise.partial_wires(qml.DepolarizingChannel, 0.01)
+
+op = qml.X('w1') # Example operation
+print(f"Error for {op}: {depol_error(op)}")
```
Modified it a bit to avoid using noise function is being evaluated.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,249 @@
+kraus_mats = [reduce(np.kron, prod, 1.0) for prod in product(pauli_mats, repeat=2)]
+def noise4(op, **kwargs):
+    probs = np.array([1 - kwargs["p"]] + [kwargs["p"] / 15] * 15).reshape((-1, 1, 1))
+    qml.QubitChannel(np.sqrt(probs) * np.array(kraus_mats), op.wires)
```
Since we don't have any explicit two-qubit error channels, I think it is useful to have one such example for readers to know the extent of the questions they can ask. Nevertheless, I have removed this. 👍

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,274 @@
+# the error operations that are inserted when a gate operation in the circuit satisfies
+# the corresponding conditional. We support the following construction of noise functions:
+#
+# 1. **Single-instruction noise functions:** To add a single-operation noise, one can use
```
We are describing the `noise` here, so no need to add `function`.

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+    qml.RY(np.pi / 8, op.wires)
+
+######################################################################
+# Finally, we can build the noise model with some required ``metadata`` for ``noise2``:
```
```suggestion
# Finally, we can build the noise model with some required ``metadata`` for ``noise1`` and ``noise2``:
```

I forgot that `p` is needed for `DepolarizingChannel`!

isaacdevlugt commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+plt.show()
+
+######################################################################
+# We can then use these for running noisy simulations as shown below:
```
```suggestion
# We can then use this for running noisy simulations as shown below:
```

isaacdevlugt reviewed (approved):
🥳 yay! Thanks a bunch @obliviateandsurrender for dealing with my persistence 😁. This looks good from my side!

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# using standard bitwise operations such as ``&``, ``|``, ``^``, or ``~``. We support
+# the following types of conditionals:
+#
+# 1. **Operation-based conditionals:** They evaluate a gate operation based on whether it
+#    is a specific type of operation or belongs to a specified set of operations. They are
+#    built using the :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
```
```suggestion
# 1. **Operation-based conditionals:** They evaluate whether a gate
#    is a specific type of operation or belongs to a specified set of operations. They are
#    built using the :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
```

alvaro-at-xanadu replied:
Less wordy

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# 1. **Operation-based conditionals:** They evaluate a gate operation based on whether it
+#    is a specific type of operation or belongs to a specified set of operations. They are
+#    built using the :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
+# 2. **Wire-based conditionals:** They evaluate a gate operation based whether if
+#    its wires are equal or belong to a specified set of wires. They are built using the
+#    :func:`~.pennylane.noise.wires_eq` and :func:`~.pennylane.noise.wires_in`.
```
I think this is what you mean? It was a bit confusing before.

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# contains the error operations that are *inserted* when a gate operation in the circuit
+# satisfies corresponding conditional. There are a few ways to construct noise functions:
+#
+# 1. **Single-instruction noise functions:** To add a single-operation noise, one can use
```
```suggestion
# 1. **Single-instruction noise functions:** To add a single-operation noise, we can use
```

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# 1. **Single-instruction noise functions:** To add a single-operation noise, one can use
+#    :func:`~pennylane.noise.partial_wires`. It performs a partial initialization of the
+#    noise operation and queues it on the ``wires`` of the gate operation.
+# 2. **User-defined noise functions:** For adding more sophisticated and custom noise,
+#    one can define their own quantum function with the signature specified above.
```
```suggestion
# 2. **User-defined noise functions:** For adding more sophisticated and custom noise,
#    we can define our own quantum function with the signature specified above.
```

alvaro-at-xanadu replied:
You started using the "we" person in the beginning. It's good to stick to it because it's also more personable.

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# 2. **User-defined noise functions:** For adding more sophisticated and custom noise,
+#    one can define their own quantum function with the signature specified above.
+#
+# For example, one can use the following for inserting a depolarization error and show
```
```suggestion
# For example, we can use the following to insert a depolarization error and show
```

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+
+def noise2(op, **kwargs):
+    for wire in op.wires:
+        qml.ThermalRelaxationError(0.1, kwargs["t1"], kwargs["t2"], kwargs["tg"], wire)
```
I don't understand what's going on. What are these "t1", "t2" etc. Maybe there should be a short explanation above.

alvaro-at-xanadu replied:
Maybe a simple way to fix this is to link "thermal relaxation errors" to the respective documentation page.

obliviateandsurrender replied:
Added the link 👍

alvaro-at-xanadu commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+noisy_circuit = qml.add_noise(ideal_circuit, noise_model)
+qml.draw_mpl(noisy_circuit)(theta, phi)
+plt.show()
+
```
The noisy gates' text is so small. I wonder if there's a way to make them bigger (if not, don't worry).

obliviateandsurrender replied:
Currently not. There's a backlog ticket on the shortcut to improve the labelling of channels that should fix it in the future.

alvaro-at-xanadu replied:
Sounds good!

alvaro-at-xanadu reviewed (commented):
Pretty good and succinct how-to! Minor copy edits.

soranjh commented on _static/demonstration_assets/noise_models/noise_model_long.jpg:
Minor, but might use "No error added" instead of "Remains Same".

Also, might just use "Damping errors added" for brevity.

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+"""
+
+######################################################################
+# Noise models are essential for understanding and describing the effects of physical errors
+# in a quantum computation, as they allow for simulating the imperfections in state evolution
+# arising from environment-based errors, state preparation routines, measurements, and more.
```
```suggestion
# Noise models are essential for describing the effects of physical errors
# in a quantum computation. They allow for simulating the imperfections in state evolution
# arising from environment-based errors, state preparation routines, measurements, and more.
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+
+######################################################################
+# In the upcoming sections, we will first cover the underlying components of
+# noise models and learn how to use them to construct one. Finally, we will use
```
```suggestion
# noise models and learn how to use them to construct desired noise models. Finally, we will use
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+#
+# 1. **Operation-based conditionals:** They evaluate whether a gate operation is a
+#    specific type of operation or belongs to a specified set of operations. They are
+#    built using the :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
```
```suggestion
#    built using :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+#    specific type of operation or belongs to a specified set of operations. They are
+#    built using the :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
+# 2. **Wire-based conditionals:** They evaluate whether a gate operation's wires are
+#    equal to or are contained in a specified set of wires. They are built using the
```
```suggestion
#    equal to or are contained in a specified set of wires. They are built using
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# Noise functions
+# ~~~~~~~~~~~~~~~
+#
+# Callables that apply noise operations are referred to as *noise functions* and have the
```
Make noise functions bold, not italic for consistency with conditionals.

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# ~~~~~~~~~~~~~~~
+#
+# Callables that apply noise operations are referred to as *noise functions* and have the
+# signature ``fn(op, **metadata) -> None``. Their definition has no return statement and
```
```suggestion
# signature ``fn(op, **metadata) -> None``. They have no return statement and
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# Noise Models
+# ~~~~~~~~~~~~
+#
+# We can now create a PennyLane :class:`~.pennylane.NoiseModel` by stitching together
+# multiple condition-callable pairs, where noisy noise operations are inserted into the
+# circuit when their corresponding given condition is satisfied. For the first pair, we
+# will use the previously constructed conditional and callable to insert a depolarization
+# error after :class:`~.pennylane.RX` gates that satisfy :math:`|\phi| < 1.0` and that
+# act on the wires :math:`\in \{0, 1\}`.
```
```suggestion
# We can now create a PennyLane :class:`~.pennylane.NoiseModel` by stitching together
# multiple condition-callable pairs, where noise operations are inserted into the
# circuit when their corresponding condition is satisfied. For the first pair, we
# will use the previously constructed conditional and callable to insert a depolarization
# error after :class:`~.pennylane.RX` gates that satisfy :math:`|\phi| < 1.0` and that
# act on the wires :math:`\in \{0, 1\}`.
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+        qml.ThermalRelaxationError(0.1, kwargs["t1"], kwargs["t2"], kwargs["tg"], wire)
+
+######################################################################
+# By default, noise operations specified by a noise function will be inserted *after*
+# the gate operation that satisfies the conditional. However, we can circumvent this by
+# manually queing the evaluated gate operation via :func:`~pennylane.apply` within the
+# function definition. For example, we can add a sandwiching constant-valued rotation
+# error for :class:`~.pennylane.Hadamard` gates on the wires :math:`\in \{0, 1\}`:
```
```suggestion
# By default, noise operations specified by a noise function will be inserted *after*
# the gate operation that satisfies the conditional. However, noise operations can be added 
# manually by queuing the evaluated gate operation via :func:`~pennylane.apply` within the
# function definition. For example, we can add a sandwiching constant-valued rotation
# error for :class:`~.pennylane.Hadamard` gates on the wires :math:`\in \{0, 1\}`:
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# Noise models provide a succinct way to describe the impact of the environment on
+# quantum computation. In PennyLane, we define such models as mapping between conditionals
+# that select the target operation and their corresponding noise operations. These
+# can be constructed with utmost flexibility, as shown above.
```
```suggestion
# can be constructed with utmost flexibility as we showed here.
```

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# Should you have any questions about using noise models in PennyLane, you can consult the
+# `noise module documentation <https://docs.pennylane.ai/en/stable/code/qml_noise.html>`__
+# or create a post on the `PennyLane Discussion Forum <https://discuss.pennylane.ai>`__.
+# You can also follow us on `X (formerly Twitter) <https://twitter.com/PennyLaneAI>`__
+# or `LinkedIn <https://www.linkedin.com/company/pennylaneai/>`__ to stay up-to-date with
+# the latest and greatest from PennyLane!
```
Fine but strange.

soranjh commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+ideal_circ_fidelity = qml.math.fidelity(ideal_circuit(theta, phi), init_dm)
+noisy_circ_fidelity = qml.math.fidelity(noisy_circuit(theta, phi), init_dm)
+
+print(f"Ideal v/s Noisy: {ideal_circ_fidelity} and {noisy_circ_fidelity}")
```
Might round to get something like
`1.0000 and 0.9017`

soranjh reviewed (approved):
Looks good, left some final small suggestions.

isaacdevlugt commented on demonstrations/tutorial_noise_models.metadata.json:
```diff
@@ -0,0 +1,35 @@
+            "id": "utkarsh_azad"
+        }
+    ],
+    "dateOfPublication": "2024-08-14T00:00:00+00:00",
+    "dateOfLastModification": "2024-08-14T00:00:00+00:00",
```
note to change!

ikurecic commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+#    equal to or are contained in a specified set of wires. They are built using
+#    :func:`~.pennylane.noise.wires_eq` and :func:`~.pennylane.noise.wires_in`.
+# 3. **Arbitrary conditionals:** Custom conditionals can be defined as a function wrapped
+#    with a :class:`~.pennylane.BooleanFn` decorator. Signature for such conditionals must
```
```suggestion
#    with a :class:`~.pennylane.BooleanFn` decorator. The signature for such conditionals must
```

ikurecic commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+print(f"Error for {op}: {depol_error(op)}")
+
+######################################################################
+# Noise Models
```
```suggestion
# Noise models
```

ikurecic commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,233 @@
+# that select the target operation and their corresponding noise operations. These
+# can be constructed with utmost flexibility as we showed here.
+#
+# Should you have any questions about using noise models in PennyLane, you can consult the
+# `noise module documentation <https://docs.pennylane.ai/en/stable/code/qml_noise.html>`__
+# or create a post on the `PennyLane Discussion Forum <https://discuss.pennylane.ai>`__.
```
Could we also add a link to a related Codebook module? https://pennylane.ai/codebook/#06-noisy-quantum-theory

ikurecic reviewed (commented):
Thanks, @obliviateandsurrender !
I left just a few minor comments, which I hope make sense to you, and we can wait until the images are ready to set a publication date. Please let me know if you have any questions.

ikurecic commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,234 @@
+#
+
+######################################################################
+# .. figure:: ../_static/demonstration_assets/noise_models/noise_model_long.jpg
+#    :align: center
+#    :width: 85%
```
This figure feels like it's a bit out context here. Is there a better place to put it?
Perhaps the OGthumbnail version of the thumbnail could be used at 70% width here if you'd like to have just any sort of image to break up the text here instead.

obliviateandsurrender replied:
It is referenced as an example in the sentence above it to show how transformation is expected to occur via a noise model.

ikurecic replied:
Let me change that to "the following figure is an example of how..." — I honestly thought that meant something further away.

ikurecic commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,234 @@
+######################################################################
+# In the upcoming sections, we will first cover the underlying components of
+# noise models and learn how to use them to construct desired noise models.
+# Finally, we will use the noise model to perform noisy simulations.
```
```suggestion
# Finally, we will use our noise model to perform noisy simulations.
```
Since you're announcing that you'll construct a desired model.

ikurecic reviewed (commented):
Thanks for the tag, @obliviateandsurrender .
I have left some comments and feedback.
Please let me know if you have any questions or tell me when you have revised the pull request. Thanks.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,234 @@
+#
+
+######################################################################
+# Conditionals
+# ~~~~~~~~~~~~
+#
+# We implement conditions as Boolean functions that accept an operation and evaluate it
+# to return a Boolean output. In PennyLane, such objects are referred to as **conditionals**.
```
Previously it was kept there but during the review process, it was discussed over and decided that details in the introductory paragraphs should be kept short - proposing a basic idea and then expanding upon them in their individual paragraphs.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,234 @@
+# in a quantum computation. They allow for simulating the imperfections in state evolution
+# arising from environment-based errors, state preparation routines, measurements, and more.
+#
+# Here, we show how to use the features provided in PennyLane's :mod:`~.pennylane.noise`
```
This is how we add references to module docs in other demos as well. I think this was not clickable in the staging due to some caching issue, but will be on the actual deployment.

obliviateandsurrender commented on demonstrations/tutorial_noise_models.py:
```diff
@@ -0,0 +1,234 @@
+# using standard bitwise operations such as ``&``, ``|``, ``^``, or ``~``. PennyLane supports
+# the following types of conditionals:
+#
+# 1. **Operation-based conditionals:** They evaluate whether a gate operation is a
+#    specific type of operation or belongs to a specified set of operations. They are
+#    built using :func:`~.pennylane.noise.op_eq` and :func:`~.pennylane.noise.op_in`.
+# 2. **Wire-based conditionals:** They evaluate whether a gate operation's wires are
+#    equal to or are contained in a specified set of wires. They are built using
+#    :func:`~.pennylane.noise.wires_eq` and :func:`~.pennylane.noise.wires_in`.
```
It was discussed and decided that this can be inferred by the readers from the documentation of those functions. In general, I don't think explaining the motivation behind function names should be important.

ikurecic commented on demonstrations/tutorial_noise_models.metadata.json:
```diff
@@ -0,0 +1,35 @@
+            "username": "whatsis"
+        }
+    ],
+    "dateOfPublication": "2024-10-07T00:00:00+00:00",
+    "dateOfLastModification": "2024-10-07T00:00:00+00:00",
```
tk
